source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/actions/task_space_actions.py

```diff
@@ -192,3 +196,388 @@ def _compute_frame_jacobian(self):
+        )
+        # Account for the offset
+        if self.cfg.body_offset is not None:
+            self._ee_pose_b[:, 0:3], self._ee_pose_b[:, 3:7] = math_utils.combine_frame_transforms(
+                self._ee_pose_b_no_offset[:, 0:3], self._ee_pose_b_no_offset[:, 3:7], self._offset_pos, self._offset_rot
+            )
+        else:
+            self._ee_pose_b[:] = self._ee_pose_b_no_offset
+
+    def _compute_ee_velocity(self):
+        """Computes the velocity of the ee frame in root frame."""
+        # Extract end-effector velocity in the world frame
+        self._ee_vel_w[:] = self._asset.data.body_vel_w[:, self._ee_body_idx, :]
+        # Compute the relative velocity in the world frame
+        relative_vel_w = self._ee_vel_w - self._asset.data.root_vel_w
+
+        # Convert ee velocities from world to root frame
+        self._ee_vel_b[:, 0:3] = math_utils.quat_rotate_inverse(self._asset.data.root_quat_w, relative_vel_w[:, 0:3])
+        self._ee_vel_b[:, 3:6] = math_utils.quat_rotate_inverse(self._asset.data.root_quat_w, relative_vel_w[:, 3:6])
+
+        # Account for the offset
+        if self.cfg.body_offset is not None:
+            # Compute offset vector in root frame
+            r_offset_b = math_utils.quat_rotate(self._ee_pose_b_no_offset[:, 3:7], self._offset_pos)
+            # Adjust the linear velocity to account for the offset
+            self._ee_vel_b[:, :3] += torch.cross(self._ee_vel_b[:, 3:], r_offset_b, dim=-1)
+            # Angular velocity is not affected by the offset
+
+    def _compute_ee_force(self):
+        """Computes the contact forces on the ee frame in root frame."""
+        # Obtain contact forces only if the contact sensor is available
+        if self._contact_sensor is not None:
+            self._contact_sensor.update(self._sim_dt)
+            self._ee_force_w[:] = self._contact_sensor.data.net_forces_w[:, 0, :]  # type: ignore
+            # Rotate forces and torques into root frame
+            self._ee_force_b[:] = math_utils.quat_rotate_inverse(self._asset.data.root_quat_w, self._ee_force_w)
+
+    def _compute_task_frame_pose(self):
+        """Computes the pose of the task frame in root frame."""
+        # Update task frame pose if task frame rigidbody is provided
+        if self._task_frame_transformer is not None and self._task_frame_pose_b is not None:
+            self._task_frame_transformer.update(self._sim_dt)
+            # Calculate the pose of the task frame in the root frame
+            self._task_frame_pose_b[:, :3], self._task_frame_pose_b[:, 3:] = math_utils.subtract_frame_transforms(
+                self._asset.data.root_pos_w,
+                self._asset.data.root_quat_w,
+                self._task_frame_transformer.data.target_pos_w[:, 0, :],
+                self._task_frame_transformer.data.target_quat_w[:, 0, :],
+            )
+
+    def _preprocess_actions(self, actions: torch.Tensor):
```

Comment:
Please add google style docstrings for arguments

Reply:
def _preprocess_actions(self, actions: torch.Tensor):
        """Pre-processes the raw actions for operational space control.

        Args:
            actions (torch.Tensor): The raw actions for operational space control. It is a tensor of
                shape (num_envs, action_dim).
        """